reactioncommerce/reaction
# No more Packages collection, implement AddressValidationRules GQL, other 3.0 stuff

Impact: **breaking**  
Type: **feature**

## Changes
Two main goals here:
- removing all remaining uses of `Packages` collection
- removing `settings` and `registry` options from all `registerPlugin` calls

Specifically:
- The address validation service rules that were previously stored in the `enabledServices` array in the `reaction-address` Packages document now have a full GraphQL API.
    - `createAddressValidationRule` mutation
    - `updateAddressValidationRule` mutation
    - `deleteAddressValidationRule` mutation
    - `addressValidationRules` query
- Includes integration tests for all of these new GraphQL mutations and queries
- When requesting address validation, these new rules (stored in `AddressValidationRules` collection) are now used instead of the `enabledServices` array.
- Move the `flatRates.enabled` setting, which allows you to turn off flat rate shipping quotes, to `isShippingRatesFulfillmentEnabled` shop setting using newer API. The setting is `true` by default.
- `Packages` are no longer upserted when creating a shop or on startup

Other:
- Add `version` to all packages set to the app version
- A couple fixes and performance improvements to the tax rate GraphQL

## Breaking changes
- If you've defined address validation rules, you'll need to recreate them with the new GraphQL mutation.
- If you want flat rate shipping disabled, you'll need to set the `isShippingRatesFulfillmentEnabled` shop setting to `false` using the `updateShopSettings` GraphQL mutation

## Testing
NOTE: Some of these changes break some aspects of Reaction Admin. A separate PR fixing Reaction Admin will be forthcoming.

- Verify API starts up without issues. On a fresh database, you should NOT see any `Packages` documents inserted. If there are existing `Packages` documents, they are left alone but no longer used.
- Test the address validation rule mutations and query using GraphQL Playground. See the integration tests for examples.
- Verify you can set and read the `isShippingRatesFulfillmentEnabled` shop setting as "admin", but not as another role or when unauthenticated.
